### PR TITLE
Connect to server

### PR DESCRIPTION
**Pull request by JochemFB**
Auto connect to server on startup
Connect button in settings
Settings UI changed
Check internet connection state

### CODE REVIEW

File: app/src/main/res/layout/activity_settings.xml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+        <TextView
+            android:id="@+id/textViewConnectionSettings"
+            android:layout_width="wrap_content"
+            android:layout_height="match_parent"
+            android:layout_weight="8"
+            android:gravity="center"
+            android:scrollHorizontally="false"
+            android:text="Tap to edit connection settings"
+
```
Comment (TimGels) [inline app/src/main/res/layout/activity_settings.xml:30]:
```suggestion
```

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -19,7 +19,9 @@
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
-        android:theme="@style/Theme.CameraSecuritySystem" >
+        android:theme="@style/Theme.CameraSecuritySystem"
```
Comment (JohannesKauffmann) [inline app/src/main/AndroidManifest.xml:22]: this changes nothing?

File: app/src/main/res/layout/activity_settings.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    android:padding="16dp"
+    android:weightSum="100">
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_gravity="center_horizontal"
+        android:layout_weight="10"
+        android:orientation="horizontal"
+
```
Comment (TimGels) [inline app/src/main/res/layout/activity_settings.xml:18]:
```suggestion
```

File: app/src/main/res/layout/activity_settings.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/settingsActivity"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+
```
Comment (TimGels) [inline app/src/main/res/layout/activity_settings.xml:8]: Is there a reason for the newlines?
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt [hunk 5]
```diff
@@ -110,4 +131,56 @@ class ServerConnection {
...
+        val port = sharedPreferences.getString(context.resources.getString(R.string.port), null)
+        val hostname =
+            sharedPreferences.getString(context.resources.getString(R.string.ip_address), null)
+
+        var credentials = arrayOf(camera_id, pwd, pwdIVByte, port, hostname)
+        if (credentials.contains(null)) {
+            return false
+        }
+        return true
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt:153]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt [hunk 6]
```diff
@@ -110,4 +131,56 @@ class ServerConnection {
 
     fun isConnected() = webSocketSession != null
 
+    private fun credentialsEntered(context: Context): Boolean {
+        var sharedPreferences =
+            context.getSharedPreferences("com.camerasecuritysystem.client", Context.MODE_PRIVATE)
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt:137]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/MainActivity.kt [hunk 7]
```diff
@@ -1,5 +1,6 @@
 package com.camerasecuritysystem.client
 
+import android.content.Context
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/MainActivity.kt:3]:
```suggestion
```
unused

File: app/src/main/kotlin/com/camerasecuritysystem/client/MainActivity.kt [hunk 8]
```diff
@@ -21,20 +22,23 @@ import kotlinx.coroutines.launch
...
     private lateinit var drawerLayout: DrawerLayout
     private lateinit var appBarConfiguration: AppBarConfiguration
 
-    private lateinit var listener : NavController.OnDestinationChangedListener
+    private lateinit var listener: NavController.OnDestinationChangedListener
 
     private var layoutManager: RecyclerView.LayoutManager? = null
     private var adapter: RecyclerView.Adapter<RecyclerAdapter.ViewHolder>? = null
 
     private var serverConnection: ServerConnection? = null
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/MainActivity.kt:39]:
```suggestion
```
unused

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt [hunk 9]
```diff
@@ -15,56 +18,72 @@ import java.nio.channels.ClosedChannelException
...
-                            val serverMessage = Json.decodeFromString<Message>(json)
-                            MessageHandler.handleMessage(serverMessage, serverConnection!!)
+                            val serverMessage: Message =
+                                Json.decodeFromString(json) //decodeFromString(json)
+                            MessageHandler.handleMessage(serverMessage, getInstance()!!)
                         }
                     }
 
+                    Log.e("WEBSOCKET", "$webSocketSession")
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt:85]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 10]
```diff
@@ -1,39 +1,108 @@
 package com.camerasecuritysystem.client
 
+import android.app.Activity
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:3]:
```suggestion
```
unused

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+            socket.connect(InetSocketAddress("8.8.8.8", 53), 1500)
+            socket.close()
+            Log.d("PING", "PING success.")
+            true
+        }catch (e: IOException){
+            Log.e("PING", "No internet connection. ${e}")
+            false
+        }
+    }
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt:24]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 12]
```diff
@@ -1,39 +1,108 @@
...
+
+        if (isNetworkAvailable == true) {
+            connectBtn.isClickable = true
+            connectBtn.isEnabled = true
+            connectBtn.backgroundTintList =
+                getColorStateList(R.color.design_default_color_primary)
+            connectBtn.text = "Connect"
+            Log.e("Button", "Enabled = ${connectBtn.isEnabled}")
+        }
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:105]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 13]
```diff
@@ -1,39 +1,108 @@
...
+        connectionLiveData = ConnectionLiveData(this)
+        connectionLiveData.observe(this, { isNetworkAvailable ->
+            Log.e("NETWORK", "Connected = $isNetworkAvailable")
+            updateUI()
+        })
+            updateUI()
+
+
+    }
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:80]:
```suggestion
            updateUI()
    }
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/ConnectionLiveData.kt [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        /*
+          If the callback was registered with registerNetworkCallback() it will be called for each network which no longer satisfies the criteria of the callback.
+          Source: https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onLost(android.net.Network)
+         */
+        override fun onLost(network: Network) {
+            Log.d(TAG, "onLost: ${network}")
+            validNetworks.remove(network)
+            checkValidNetworks()
+        }
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/ConnectionLiveData.kt:79]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 15]
```diff
@@ -1,39 +1,108 @@
...
+        }
+
+        connectionLiveData = ConnectionLiveData(this)
+        connectionLiveData.observe(this, { isNetworkAvailable ->
+            Log.e("NETWORK", "Connected = $isNetworkAvailable")
+            updateUI()
+        })
+            updateUI()
+
+
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:79]:
```suggestion
        updateUI()
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 16]
```diff
@@ -1,39 +1,108 @@
...
+            connectBtn.isClickable = false
+            connectBtn.isEnabled = false
+            connectBtn.backgroundTintList =
+                getColorStateList(R.color.cardview_dark_background)
+
+            connectBtn.text = "NO Signal"
+            Log.e("Button", "Enabled = ${connectBtn.isEnabled}")
+        }
+
+        if (isNetworkAvailable == true) {
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:97]: maybe else?

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 17]
```diff
@@ -46,30 +115,43 @@ class SettingsActivity : AppCompatActivity(),
         return false
     }
 
-    override fun applyTexts(cameraID: String?, port: String?, ipAddress: String?, password: String?) {
+    override fun applyTexts(
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:118]: This doesnt change anything and will conflict...

File: app/src/main/kotlin/com/camerasecuritysystem/client/ConnectionLiveData.kt [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+                            checkValidNetworks()
+                        }
+                    }
+                }
+            }
+        }
+
+
+
+        /*
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/ConnectionLiveData.kt:70]:
```suggestion
        }

        /*
```

File: app/src/main/res/layout/activity_settings.xml [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_gravity="center_horizontal"
+        android:layout_weight="10"
+        android:orientation="horizontal"
+
+        android:weightSum="10">
+
```
Comment (TimGels) [inline app/src/main/res/layout/activity_settings.xml:20]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt [hunk 20]
```diff
@@ -15,56 +18,72 @@ import java.nio.channels.ClosedChannelException
...
                             val json = String(frame.readBytes())
-                            val serverMessage = Json.decodeFromString<Message>(json)
-                            MessageHandler.handleMessage(serverMessage, serverConnection!!)
+                            val serverMessage: Message =
+                                Json.decodeFromString(json) //decodeFromString(json)
+                            MessageHandler.handleMessage(serverMessage, getInstance()!!)
                         }
                     }
 
+                    Log.e("WEBSOCKET", "$webSocketSession")
```
Comment (JohannesKauffmann) [inline app/src/main/kotlin/com/camerasecuritysystem/client/models/ServerConnection.kt:84]: Maybe remove logging?

File: app/src/main/kotlin/com/camerasecuritysystem/client/MainActivity.kt [hunk 21]
```diff
@@ -43,27 +47,38 @@ class MainActivity : AppCompatActivity() {
...
+            if (isNetworkAvailable) {
+                GlobalScope.launch {
+                    ServerConnection.getInstance().connectIfPossible(context)
+                }
             }
-        }
 
-        listener = NavController.OnDestinationChangedListener{ _, destination, _ ->
+        })
+
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/MainActivity.kt:63]:
```suggestion
```

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt [hunk 22]
```diff
@@ -1,39 +1,108 @@
...
+            //Try to setup a connection
+            GlobalScope.launch {
+                try {
+                    val connection = ServerConnection.getInstance()
+                    if (!connection.isConnected()){
+                        connection.connectIfPossible(context)
+
+                        Log.e("Connection", "Verbinden mogelijk: ${connection.connectIfPossible(context)}")
+                    }else{
+                        Log.e("Connection", "Ik ben al verbonden")
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:63]: We use English as our main language
```suggestion
                        Log.e("Connection", "Connecting is possible: ${connection.connectIfPossible(context)}")
                    }else{
                        Log.e("Connection", "Already connected")
```

**Review by JohannesKauffmann — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+import javax.net.SocketFactory
+
+
+object DoesNetworkHaveInternet {
+    // Make sure to execute this on a background thread.
+    fun execute(socketFactory: SocketFactory): Boolean {
+        return try{
+            Log.d("PING", "PINGING google.")
+            val socket = socketFactory.createSocket() ?: throw IOException("Socket is null.")
+            socket.connect(InetSocketAddress("8.8.8.8", 53), 1500)
```
Comment (TimGels) [inline app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt:15]: We still want to ping google? Is it possible to ping the IP address of the server?

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt (hunk 22 above)
Comment (JohannesKauffmann) [reply app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt:63]: Or just remove logs?

File: app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt (hunk 23 above)
Comment (JochemFB) [reply app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt:15]: Noted. I made a ToDo card in Trello

Comment (TimGels) [reply app/src/main/kotlin/com/camerasecuritysystem/client/models/DoesNetworkHaveInternet.kt:15]: Is on the roadmap

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt (hunk 17 above)
Comment (JochemFB) [reply app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt]: Removed the change

File: app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt (hunk 16 above)
Comment (JochemFB) [reply app/src/main/kotlin/com/camerasecuritysystem/client/SettingsActivity.kt]: Changed to else if

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (JochemFB) [reply app/src/main/AndroidManifest.xml:22]: fixed

**Review by JohannesKauffmann — APPROVED** [review]

**Review by TimGels — APPROVED** [review]

**Pull request merged**